doc/arch/adr-002-delete-unpublished-documents-from-index.md

```diff
@@ -0,0 +1,90 @@
+# Decision record: delete documents from the govuk search index when an unpublishing message is received
+
+## Context
+
+We are in the process of creating a new 'govuk' index to replace the existing
+mainstream, government and detailed indexes. The new index will be populated by
+rummager using publishing and unpublishing event messages pulled from a queue
+controlled by the publishing API.
+
+Rummager cannot rely on messages from the queue being processed in the order
+that they were generated. This means that either rummager should handle edge
+cases such as these or we will need to accept the consequences:
+
+- A document is published twice in quick succession. The second publishing event
+  arrives in rummager first. The later, stale message should not overwrite the
+  more recent data in the index.
+- A document is unpublished and then immediately republished. The republishing
+  message is processed first. The later, stale unpublishing message should be
+  ignored and the published document should remain in the search index.
+- A document is published and then immediately unpublished. The unpublishing
+  event is processed first. The later, stale publishing message should not
+  re-add the document to the search index. It should remain unpublished
+
+## Decision
+
+### Use Elasticsearch's external versioning
+
+Elasticsearch supports [external versioning][es-versioning] for documents. This
+means that each document includes a version number from an external system. We
+will use the publishing event number from the publishing API.
+
+When the version number is set, Elasticsearch uses optimistic locking to ensure
+that a document is never updated or deleted if the new version number is less
+than or equal to the existing version number.
+
+This will take care of some of the edge cases:
+
+- If two publishing events arrive out of order, the second stale event will be
+  ignored because the version number is lower. Result: the document in the index
+  ends up in the expected state, which is the most recent version.
+- If a republishing message arrives before an earlier unpublishing message, the
+  unpublishing will be ignored because its version number is lower. Result: the
+  document remains in the index as expected.
+
+[es-versioning]: https://www.elastic.co/blog/elasticsearch-versioning-support
+
+### Delete unpublished documents
```

Comment:
These seem like options for the last edge case rather than `decisions` which they are nested under.